grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,150 @@ class PrepaymentInvoiceService {
+                invoice.addToInvoiceItems(invoiceItem)
+                if (inverseItem) {
+                    invoice.addToInvoiceItems(inverseItem)
                 }
             }
         }
 
         order.invoiceableAdjustments.each { OrderAdjustment orderAdjustment ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderAdjustment(orderAdjustment)
+            InvoiceItem invoiceItem = createFromOrderAdjustment(orderAdjustment)
+            InvoiceItem inverseItem = createInverseItemForOrderAdjustment(orderAdjustment)
             invoice.addToInvoiceItems(invoiceItem)
+            invoice.addToInvoiceItems(inverseItem)
         }
 
         return invoice.save()
     }
 
-    Invoice generatePrepaymentInvoice(Order order) {
-        if (order.orderItems.any { it.hasInvoices } || order.orderAdjustments.any { it.hasInvoices }) {
-            throw new Exception("Some order items or order adjustments for this order already have been invoiced")
+    InvoiceItem createFromOrderItem(OrderItem orderItem) {
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            InvoiceItem invoiceItem = new InvoiceItem(
+                    quantity: 0,
+                    product: orderItem.product,
+                    glAccount: orderItem.glAccount ?: orderItem.product.glAccount,
+                    budgetCode: orderItem?.budgetCode,
+                    quantityUom: orderItem?.quantityUom,
+                    quantityPerUom: orderItem.quantityPerUom ?: 1,
+                    unitPrice: orderItem.unitPrice,
+                    // canceled invoice items amount is 0
+                    amount: 0,
+                    invoiceItemType: InvoiceItemType.REGULAR
+            )
+            invoiceItem.addToOrderItems(orderItem)
+            return invoiceItem
         }
 
-        Invoice invoice = invoiceService.createFromOrder(order)
-        invoice.invoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
-        invoiceService.createOrUpdateVendorInvoiceNumber(invoice, order.orderNumber + Constants.PREPAYMENT_INVOICE_SUFFIX)
+        InvoiceItem invoiceItem = new InvoiceItem(
+                quantity: orderItem.quantity,
+                product: orderItem.product,
+                glAccount: orderItem.glAccount ?: orderItem.product.glAccount,
+                budgetCode: orderItem?.budgetCode,
+                quantityUom: orderItem?.quantityUom,
+                quantityPerUom: orderItem.quantityPerUom ?: 1,
+                unitPrice: orderItem.unitPrice,
+                // For non-canceled order items invoice item amount is equal to (order item quantity * unit price)
+                amount: (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0),
+                invoiceItemType: InvoiceItemType.REGULAR
+        )
+        invoiceItem.addToOrderItems(orderItem)
+        return invoiceItem
+    }
```

Comment:
You could simplify this down to something like this if you want:
```
  private InvoiceItem createFromOrderItem(OrderItem orderItem) {
      // For non-canceled order items invoice item amount is equal to (order item quantity * unit price)
      int amount = orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED
              ? 0 : (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0)

      InvoiceItem invoiceItem = new InvoiceItem(
              quantity: orderItem.quantity,
              product: orderItem.product,
              glAccount: orderItem.glAccount ?: orderItem.product.glAccount,
              budgetCode: orderItem?.budgetCode,
              quantityUom: orderItem?.quantityUom,
              quantityPerUom: orderItem.quantityPerUom ?: 1,
              unitPrice: orderItem.unitPrice,
              amount: amount,
              invoiceItemType: InvoiceItemType.REGULAR
      )
      invoiceItem.addToOrderItems(orderItem)
      return invoiceItem
  }
```
I know you probably did it this way intentionally to keep the cancelled and non-cancelled flow separate but it's probably fine to condense it at this point since it's just one field that's different. Up to you though.

Reply:
Yeah I think I am going to leave it as is for now (could be adjusted together with tests coverage ticket)